### DISCUSSION TITLE
Define a daily partition which overwrites previous month

### DISCUSSION THREAD

**Discussion by cody-scott** [dagster-io/dagster#19111] [Q&A] [answered]
Hello, looking to define a partition that runs daily, but overwrites from the partition date back a month to address late arriving data.

I think something along the lines of `DailyPartitionsDefinition` with an `window` size argument is what i'm thinking. The primary downside I see to this is you mutate a historical partition slice in a run, which overlaps other partitions; certainly doesn't *feel* right, if that makes sense.

Other option is to forgo partitions and the IO manager and manage it ourselves.

How are others dealing with late arriving data in dagster?

**Comment by clairelin135** [comment, marked as answer]
Hi @cody-scott thanks for the detail here.

It isn't possible for a partition to overlap with other partitions. The idea here is that partitioning allows datasets to be split into manageable chunks and that executing a given partition will only load the data that's needed. This helps to structure data pipelines in a way that limits risk and reduces cost.

I would recommend that each partition only overwrites data for the corresponding date to take advantage of these benefits.

One possible approach to handle late arriving data: Say you have a daily schedule that materializes today's partition. You could customize that schedule to also detect which older partitions received new data today, and yield a run request that refreshes each updated partition.

**Reply by cody-scott** [reply]
Ok, that is in line with how i felt too; glad it tracks! The business unit is essentially asking to refresh (today - a year) each day. Personal preference would have been to deal with it in something like dbt, but its based on an external R script unfortunately.

The schedule option was something i was considering as well, basically just trigging N number of runs based on the schedule. Will this be triggered as a backfill? The single run backfill is great to materialize everything at once, not sure if that would be the case with multiple run requests, my first guess it would generate N runs for N partitions. Does that align with your perspective?

My last option is to forgo the IO and just manage the logic within the asset function, which is not preferred, but functional.

Thanks for the help!

**Reply by clairelin135** [reply]
@cody-scott If you yield a standard run request per partition, i.e:
```
return RunRequest(partition_key=...)
```
that will kick off a run per partition.

[Here's an example](https://github.com/dagster-io/dagster/discussions/14622) of how you can kick off a range of partitions per run request. You'd need to write logic to find the contiguous ranges within the set of partitions you want to launch, then return a run request per range.

**Reply by cody-scott** [reply]
That might be what I'm looking for! I'll try it out tomorrow! Thanks again!

**Reply by cody-scott** [reply]
Hi Claire, one last question.

I have it working successfully, I'm just a bit fuzzy on the testing piece.

Any recommendations on how to set this up for testing? I believe i can generate a `Definition` in the test, then call it from there, but I think that would skip the function itself and whats its generating. 

Imagining the `build_schedule_context` is along the lines of what's needed, but not clear on how to pass that into or trigger a schedule to evaluate.

**Reply by cody-scott** [reply]
this is what i have so far as well.

```python
def test_windowed_partitions_schedule():
    expected = {
        "dagster/asset_partition_range_start": "2024-01-01-0500",
        "dagster/asset_partition_range_end": "2024-01-01-0500",
        "dagster/schedule_name": "partition_sample2_job_schedule",
    }

    defs = Definitions(
        assets=data["assets"],
        jobs=data["jobs"],
        schedules=data["schedules"],
        resources={"io_manager": InMemoryIOManager()},
    )

    schedule = defs.get_schedule_def("partition_sample2_job_schedule")
    context = build_schedule_context(
        scheduled_execution_time=pendulum.datetime(2024, 1, 2, tz="America/Toronto")
    )
    result = schedule.evaluate_tick(context)
    assert result.run_requests is not None
    run_requests = result.run_requests
    assert len(run_requests) == 1
    assert run_requests[0].tags == expected


```
